File: native/Sources/IsometryCore/Domain/Entities/Node.swift
import Foundation

/// Primary data entity representing a card in Isometry
public struct Node: Codable, Sendable, Identifiable, Hashable {
    // MARK: - Core Identity
    public let id: String
    public var nodeType: String
    public var name: String
    public var content: String?
    public var summary: String?

    // MARK: - LATCH: Location
    public var latitude: Double?
    public var longitude: Double?
    public var locationName: String?
    public var locationAddress: String?

    // MARK: - LATCH: Time
    public var createdAt: Date
    public var modifiedAt: Date
    public var dueAt: Date?
    public var completedAt: Date?
    public var eventStart: Date?
    public var eventEnd: Date?

    // MARK: - LATCH: Category
    public var folder: String?
    public var tags: [String]
    public var status: String?

    // MARK: - LATCH: Hierarchy
    public var priority: Int
    public var importance: Int
    public var sortOrder: Int

    // MARK: - Metadata
    public var source: String?
    public var sourceId: String?
    public var sourceUrl: String?
    public var deletedAt: Date?
    public var version: Int

    // MARK: - Sync
    public var syncVersion: Int
    public var lastSyncedAt: Date?
    public var conflictResolvedAt: Date?

    // MARK: - Initialization

    public init(
        id: String = UUID().uuidString,
        nodeType: String = "note",
        name: String,
        content: String? = nil,
        summary: String? = nil,
        latitude: Double? = nil,
        longitude: Double? = nil,
        locationName: String? = nil,
        locationAddress: String? = nil,
        createdAt: Date = Date(),
        modifiedAt: Date = Date(),
        dueAt: Date? = nil,
        completedAt: Date? = nil,
        eventStart: Date? = nil,
        eventEnd: Date? = nil,
        folder: String? = nil,
        tags: [String] = [],
        status: String? = nil,
        priority: Int = 0,
        importance: Int = 0,
        sortOrder: Int = 0,
        source: String? = nil,
        sourceId: String? = nil,
        sourceUrl: String? = nil,
        deletedAt: Date? = nil,
        version: Int = 1,
        syncVersion: Int = 0,
        lastSyncedAt: Date? = nil,
        conflictResolvedAt: Date? = nil
    ) {
        self.id = id
        self.nodeType = nodeType
        self.name = name
        self.content = content
        self.summary = summary
        self.latitude = latitude
        self.longitude = longitude
        self.locationName = locationName
        self.locationAddress = locationAddress
        self.createdAt = createdAt
        self.modifiedAt = modifiedAt
        self.dueAt = dueAt
        self.completedAt = completedAt
        self.eventStart = eventStart
        self.eventEnd = eventEnd
        self.folder = folder
        self.tags = tags
        self.status = status
        self.priority = priority
        self.importance = importance
        self.sortOrder = sortOrder
        self.source = source
        self.sourceId = sourceId
        self.sourceUrl = sourceUrl
        self.deletedAt = deletedAt
        self.version = version
        self.syncVersion = syncVersion
        self.lastSyncedAt = lastSyncedAt
        self.conflictResolvedAt = conflictResolvedAt
    }
}

// MARK: - Coding Keys for JSON Serialization

extension Node {
    public enum CodingKeys: String, CodingKey {
        case id
        case nodeType = "node_type"
        case name
        case content
        case summary
        case latitude
        case longitude
        case locationName = "location_name"
        case locationAddress = "location_address"
        case createdAt = "created_at"
        case modifiedAt = "modified_at"
        case dueAt = "due_at"
        case completedAt = "completed_at"
        case eventStart = "event_start"
        case eventEnd = "event_end"
        case folder
        case tags
        case status
        case priority
        case importance
        case sortOrder = "sort_order"
        case source
        case sourceId = "source_id"
        case sourceUrl = "source_url"
        case deletedAt = "deleted_at"
        case version
        case syncVersion = "sync_version"
        case lastSyncedAt = "last_synced_at"
        case conflictResolvedAt = "conflict_resolved_at"
    }
}

// MARK: - Computed Properties

extension Node {
    /// Whether this node has been soft-deleted
    public var isDeleted: Bool {
        deletedAt != nil
    }

    /// Whether this node has location data
    public var hasLocation: Bool {
        latitude != nil && longitude != nil
    }

    /// Whether this node is overdue
    public var isOverdue: Bool {
        guard let dueAt, completedAt == nil else { return false }
        return dueAt < Date()
    }
}

File: src/types/node.ts
// Node (Card) types
//
// @deprecated This file is being replaced by card.ts in Phase 84.
// Use Card, Connection, and related types from './card' instead.
// Migration: node_type -> card_type, edges -> connections

// Re-export new types for gradual migration
export type { Card, CardType, Connection } from './card';
export { isNote, isPerson, isEvent, isResource, rowToCard, rowToConnection } from './card';

/**
 * @deprecated Use CardType from './card' instead.
 * Maps: task/project/notebook -> 'note', contact -> 'person'
 */
export type NodeType = 'note' | 'task' | 'contact' | 'event' | 'project' | 'resource' | 'notebook';

/**
 * @deprecated Use TaskStatus from './card' instead.
 */
export type TaskStatus = 'active' | 'pending' | 'completed' | 'archived';

/**
 * @deprecated Use Connection from './card' instead.
 * The new model uses label (string) instead of edge_type (enum).
 */
export type EdgeType = 'LINK' | 'NEST' | 'SEQUENCE' | 'AFFINITY';

/**
 * @deprecated Use Card from './card' instead.
 * This interface will be removed in a future version.
 */
export interface Node {
  id: string;
  nodeType: NodeType;
  name: string;
  content: string | null;
  summary: string | null;

  // LATCH: Location
  latitude: number | null;
  longitude: number | null;
  locationName: string | null;
  /** @deprecated Removed in cards model */
  locationAddress: string | null;

  // LATCH: Time
  createdAt: string;
  modifiedAt: string;
  dueAt: string | null;
  completedAt: string | null;
  eventStart: string | null;
  eventEnd: string | null;

  // LATCH: Category
  folder: string | null;
  tags: string[];
  status: TaskStatus | null;

  // LATCH: Hierarchy
  priority: number;
  /** @deprecated Removed in cards model (redundant with priority) */
  importance: number;
  sortOrder: number;

  // Metadata
  source: string | null;
  sourceId: string | null;
  /** @deprecated Use url field on ResourceCard instead */
  sourceUrl: string | null;
  deletedAt: string | null;
  version: number;
}

/**
 * @deprecated Use Connection from './card' instead.
 */
export interface Edge {
  id: string;
  edgeType: EdgeType;
  sourceId: string;
  targetId: string;
  label: string | null;
  weight: number;
  /** @deprecated Removed in connections model (direction is view concern) */
  directed: boolean;
  /** @deprecated Removed - use sort_order on cards */
  sequenceOrder: number | null;
  /** @deprecated Removed - use via_card */
  channel: string | null;
  /** @deprecated Removed - use via_card's created_at */
  timestamp: string | null;
  /** @deprecated Removed - use via_card's name */
  subject: string | null;
  createdAt: string;
}

/**
 * @deprecated Use rowToCard from './card' instead.
 */
export function rowToNode(row: Record<string, unknown>): Node {
  return {
    id: row.id as string,
    nodeType: row.node_type as NodeType,
    name: row.name as string,
    content: row.content as string | null,
    summary: row.summary as string | null,
    latitude: row.latitude as number | null,
    longitude: row.longitude as number | null,
    locationName: row.location_name as string | null,
    locationAddress: row.location_address as string | null,
    createdAt: row.created_at as string,
    modifiedAt: row.modified_at as string,
    dueAt: row.due_at as string | null,
    completedAt: row.completed_at as string | null,
    eventStart: row.event_start as string | null,
    eventEnd: row.event_end as string | null,
    folder: row.folder as string | null,
    tags: row.tags ? JSON.parse(row.tags as string) : [],
    status: row.status as TaskStatus | null,
    priority: (row.priority as number) ?? 0,
    importance: (row.importance as number) ?? 0,
    sortOrder: (row.sort_order as number) ?? 0,
    source: row.source as string | null,
    sourceId: row.source_id as string | null,
    sourceUrl: row.source_url as string | null,
    deletedAt: row.deleted_at as string | null,
    version: (row.version as number) ?? 1,
  };
}

// ============================================
// Migration Helpers
// ============================================

import type { Card, CardType } from './card';

/**
 * Convert legacy Node to new Card format.
 * Use this during migration period.
 */
export function nodeToCard(node: Node): Card {
  // Map old node_type to new card_type
  const typeMap: Record<NodeType, CardType> = {
    note: 'note',
    task: 'note',
    project: 'note',
    notebook: 'note',
    contact: 'person',
    event: 'event',
    resource: 'resource',
  };

  const cardType = typeMap[node.nodeType] || 'note';

  const base = {
    id: node.id,
    name: node.name,
    content: node.content,
    summary: node.summary,
    latitude: node.latitude,
    longitude: node.longitude,
    locationName: node.locationName,
    createdAt: node.createdAt,
    modifiedAt: node.modifiedAt,
    dueAt: node.dueAt,
    completedAt: node.completedAt,
    eventStart: node.eventStart,
    eventEnd: node.eventEnd,
    folder: node.folder,
    tags: node.tags,
    status: node.status,
    priority: node.priority,
    sortOrder: node.sortOrder,
    source: node.source,
    sourceId: node.sourceId,
    deletedAt: node.deletedAt,
    version: node.version,
    syncStatus: 'pending' as const,
  };

  switch (cardType) {
    case 'person':
      return { ...base, cardType: 'person', isCollective: false, url: null, mimeType: null };
    case 'event':
      return { ...base, cardType: 'event', isCollective: false, url: null, mimeType: null };
    case 'resource':
      return { ...base, cardType: 'resource', url: node.sourceUrl, mimeType: null, isCollective: false };
    default:
      return { ...base, cardType: 'note', isCollective: false, url: null, mimeType: null };
  }
}
